### ISSUE TITLE
'Product of Nibbles' level needs greater test coverage.

### ISSUE THREAD

**Issue by altrag** [Turing-Complete-Game/Suggestions-and-Issues#158]
### What happened?

The current set of tests is fixed and small enough to be implemented via a moderately large but otherwise straight-forward logic decoder.

Proposed solutions (in order of my preference, not necessarily practicality!) are:
1) Test all 256 possibilities.  I don't think this is viable at the moment (would take to long to run), but future game speed improvements and/or enabling the FF button on earlier levels could allow it.

2) Use random test cases (possibly retaining a subset of fixed tests if desired, as is done on several other levels).  Almost as good as testing all combinations while still preventing cheated solutions.

3) But if we want to stick with a fixed non-randomized list of test case inputs, I've run a (nowhere near exhaustive) set of trials and the top 16 inputs, in order of increasing gate/buffer count (according to my particular logic solver's capabilities) are:
5*7, 3*14, 9*11, 14*10, 15*9, 15*8, 10*12, 10*13, 11*9, 3*3, 5*5, 1*15, 14*3, 10*10, 12*7, 11*6

Adding the full set of 16 increases the expected nand count by ~50% and almost doubles the number of buffers.  I'm not entirely sure that would be sufficient to prevent really stuffing the components in there, but it would certainly make doing so a lot more challenging!

### Version

Development

### Operating System

Windows

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by EnderShadow** [comment]
Github turned your asterisks into atalic markers. You might want to escape them.

**Comment by altrag** [comment]
Well that's annoying.  Thanks for the notice, should be fixed now.

**Comment by scottanderson** [comment]
The best solution to 4x4-bit multiply is a 256-entry LUT of the product. Any gaps in the fixed test case coverage will translate to nand savings in that LUT. This could be fixed by running all 256 test cases, or by using truly random inputs. Since byte adder already has 256 test cases, it seems like this one should just test all possible inputs.

**Comment by Stuffe** [comment]
I think this was fixed? If not please upload an example cheese solution here: https://turingcomplete.game/report_bug

**Comment by altrag** [comment]
No, my 33/4 still completes the level as of 0.367.  I'll upload the save folder.

**Comment by altrag** [comment]
My current 33/4 solution no longer works as of 0.380.  Haven't yet checked if a new LUT-based solution (short of full LUT, of course) would suffice.  Will report back if I discover such.
